Apromore-ProcessMining-Collection/prom-bpmn/src/main/java/org/apromore/processmining/models/graphbased/directed/bpmn/ModelChecker.java

```diff
@@ -0,0 +1,212 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.apromore.processmining.models.graphbased.directed.bpmn;
+
+import org.apromore.processmining.models.graphbased.directed.bpmn.elements.Activity;
+import org.apromore.processmining.models.graphbased.directed.bpmn.elements.Event;
+import org.apromore.processmining.models.graphbased.directed.bpmn.elements.Flow;
+import org.apromore.processmining.models.graphbased.directed.bpmn.elements.Gateway;
+import org.springframework.util.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class ModelChecker {
+    private static final String DISCONNECTED_ARC_MSG = "The model has disconnected arcs";
+    private static final String EMPTY_MODEL_MSG = "The model is empty";
+    private static final String POOLS_NOT_SUPPORTED_MSG = "There are pools in the model. Pools are not yet supported";
+    private static final String SELF_LOOP_MSG_FORMAT = "The element %s has a self-loop";
+    private static final String TASK_MULTIPLE_OUTGOING_ARCS_MSG_FORMAT = "The task %s has more than one outgoing arc";
+    private static final String TASK_MULTIPLE_INCOMING_ARCS_MSG_FORMAT = "The task %s has more than one incoming arc";
+    private static final String TASK_MISSING_ARCS_MSG_FORMAT = "The task %s has missing incoming or outgoing arcs";
+    private static final String GATE_MISSING_ARCS_MSG_FORMAT = "The gateway %s has missing incoming or outgoing arcs";
+    private static final String START_MULTIPLE_OUTGOING_ARCS_MSG = "The Start Event has more than one outgoing arc";
+    private static final String START_NO_OUTGOING_ARC_MSG = "The Start Event has a missing outgoing arc";
+    private static final String START_INCOMING_ARCS_MSG = "The Start Event has incoming arc(s)";
+    private static final String END_MULTIPLE_INCOMING_ARCS_MSG = "The End Event has more than one incoming arc";
+    private static final String END_NO_INCOMING_ARC_MSG = "The End Event has a missing incoming arc";
+    private static final String END_OUTGOING_ARCS_MSG = "The End Event has outgoing arc(s)";
+
+
+    /**
+     * Finds errors in a model.
+     * @param bpmnDiagram the model to find errors in.
+     * @param allowPools true if pools are supported.
+     * @return a object containing the errors in the model.
+     */
+    public ModelCheckResult checkModel(BPMNDiagram bpmnDiagram, boolean allowPools) {
```

Comment:
As discussed with @brucenguyen1 we can move this allowPools parameter to the ModelChecker constructor.